src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1475 @@
+            public void visitVarDef(JCVariableDecl tree) {
+                final Lint lintPrev = this.lint;
+                this.lint = this.lint.augment(tree.sym);
+                try {
+                    // Track warning suppression
+                    if (!this.lint.isEnabled(Lint.LintCategory.THIS_ESCAPE))
+                        ThisEscapeAnalyzer.this.suppressed.add(tree.sym);
+
+                    // Recurse
+                    super.visitVarDef(tree);
+                } finally {
+                    this.lint = lintPrev;
+                }
+            }
+
+            @Override
+            public void visitMethodDef(JCMethodDecl tree) {
+                final Lint lintPrev = this.lint;
+                this.lint = this.lint.augment(tree.sym);
+                try {
+                    // Track warning suppression
+                    if (!this.lint.isEnabled(Lint.LintCategory.THIS_ESCAPE))
+                        ThisEscapeAnalyzer.this.suppressed.add(tree.sym);
+
+                    // Determine if this is a constructor we should analyze
+                    final boolean analyzable = this.currentClassIsExternallyExtendable() &&
+                        TreeInfo.isConstructor(tree) &&
+                        !tree.sym.isPrivate() &&
+                        !ThisEscapeAnalyzer.this.suppressed.contains(tree.sym);
+
+                    // Determine if this method is "invokable" in an analysis (can't be overridden)
+                    final boolean invokable = !this.currentClassIsExternallyExtendable() ||
+                        TreeInfo.isConstructor(tree) ||
+                        (tree.mods.flags & (Flags.STATIC | Flags.PRIVATE | Flags.FINAL)) != 0;
+
+                    // Add method or constructor to map
+                    final MethodInfo info = new MethodInfo(this.currentClass, tree, analyzable, invokable);
+                    ThisEscapeAnalyzer.this.methodMap.put(tree.sym, info);
+
+                    // Recurse
+                    super.visitMethodDef(tree);
+                } finally {
+                    this.lint = lintPrev;
+                }
+            }
+
+            // Determines if the current class could be extended in some external compilation unit
+            private boolean currentClassIsExternallyExtendable() {
+                return !this.currentClass.sym.isFinal() &&
+                  !(this.currentClass.sym.isSealed() && this.currentClass.permitting.isEmpty()) &&
+                  !(this.currentClass.sym.owner.kind == MTH) &&
+                  !this.privateOuter;
+            }
+        }.scan(tree);
+
+        // TODO: eliminate sealed classes where all permitted subclasses are in this compilation unit
+
+        // Now analyze all of the analyzable constructors we found
+        for (Map.Entry<Symbol, MethodInfo> entry : this.methodMap.entrySet()) {
+
+            // We are looking for analyzable constructors only
+            final Symbol sym = entry.getKey();
```

Comment:
Thanks, you're right, the map keys are not used here. I'll clean up the loop.

But we still need a map, not a set, because the map is used elsewhere.